src/Microsoft.AspNetCore.Mvc.TagHelpers/LinkTagHelper.cs

```diff
@@ -402,38 +398,60 @@ private void EnsureFileVersionProvider()
             }
         }
 
-        private void BuildLinkTag(TagHelperAttributeList attributes, TagHelperContent builder)
+        private void BuildLinkTag(string href, TagHelperAttributeList attributes, TagHelperContent builder)
         {
             builder.AppendHtml("<link ");
 
+            var addHref = true;
+
             // Perf: Avoid allocating enumerator
             for (var i = 0; i < attributes.Count; i++)
             {
                 var attribute = attributes[i];
-                var attributeValue = attribute.Value;
-                if (AppendVersion == true &&
-                    string.Equals(attribute.Name, HrefAttributeName, StringComparison.OrdinalIgnoreCase))
+
+                if (string.Equals(attribute.Name, HrefAttributeName, StringComparison.OrdinalIgnoreCase))
                 {
-                    // "href" values come from bound attributes and globbing. So anything but a non-null string is
-                    // unexpected but could happen if another helper targeting the same element does something odd.
-                    // Pass through existing value in that case.
-                    var attributeStringValue = attributeValue as string;
-                    if (attributeStringValue != null)
-                    {
-                        attributeValue = _fileVersionProvider.AddFileVersionToPath(attributeStringValue);
-                    }
+                    addHref = false;
+
+                    AppendEncodedVersionedHref(attribute.Name, href, builder);
                 }
+                else
+                {
+                    AppendAttribute(attribute.Name, attribute.Value, builder);
+                }
+            }
 
-                builder
-                    .AppendHtml(attribute.Name)
-                    .AppendHtml("=\"")
-                    .Append(HtmlEncoder, attributeValue)
-                    .AppendHtml("\" ");
+            // May have no "href" attribute in the dictionary e.g. if Href and HrefInclude were not bound.
+            if (addHref)
+            {
+                AppendEncodedVersionedHref(HrefAttributeName, href, builder);
             }
 
             builder.AppendHtml("/>");
         }
 
+        private void AppendEncodedVersionedHref(string hrefName, string hrefValue, TagHelperContent builder)
+        {
+            if (AppendVersion == true)
+            {
+                hrefValue = _fileVersionProvider.AddFileVersionToPath(hrefValue);
+            }
+
+            // attribute.Key ("src") does not need to be JavaScript-encoded.
+            var encodedValue = JavaScriptEncoder.Encode(hrefValue);
```

Comment:
Remove this line. The `<link>` element's `href` attribute value never needs to be JavaScript encoded.